docs/development/release-process.rst

```diff
@@ -0,0 +1,96 @@
+that ``master`` is always in a releasable state. It is acceptable to merge
+PRs that partially implement a new feature, but only if the partially
+implemented version is usable in that state (for example, with reduced
+functionality or disabled by default). In the case where a merged PR is found
+to need extra work before being released, the release manager always has the
+option to back out the partial change prior to a release. The PR can then be
+reworked and resubmitted for the next release.
+
+
+Deprecation Policy
+==================
+
+Any change to pip that removes or significantly alters user-visible behaviour
+that is described in the pip documentation will be deprecated for a minimum of
+one released version before the change occurs. Deprecation will take the form of
+a warning being issued by pip when the feature is used. Longer deprecation
+periods, or deprecation warnings for behaviour changes that would not normally
+be covered by this policy, are also possible depending on circumstances, but
+this is at the discretion of the pip developers.
+
+Note that the documentation is the sole reference for what counts as agreed
+behaviour. If something isn't explicitly mentioned in the documentation, it can
+be changed without warning, or any deprecation period, in a pip release.
+However, we are aware that the documentation isn't always complete - PRs that
+document existing behaviour with the intention of covering that behaviour with
+the above deprecation process are always acceptable, and will be considered on
+their merits.
+
+
+Release Ops
+===========
+
+Doing a new release
+-------------------
+
+#. On the current pip ``master`` branch, generate a new ``AUTHORS.txt`` by
+   running ``invoke generate.authors`` and commit the results.
+#. On the current pip ``master`` branch, make a new commit which bumps the
+   version in ``pip/__init__.py`` to the release version and adjust the
+   ``CHANGES.txt`` file to reflect the current date. The release version should
+   follow a YY.N scheme, where YY is the two digit year, and N is the Nth
+   release within that year.
+#. On the current pip ``master`` branch, generate a new ``NEWS.rst`` by running
+   ``invoke generate.news`` and commit the results.
+#. Create a signed tag of the ``master`` branch of the form ``X.Y.Z`` using the
+   command ``git tag -s X.Y.Z``.
+#. Checkout the tag using ``git checkout X.Y.Z`` and create the distribution
+   files using ``python setup.py sdist bdist_wheel``.
+#. Upload the distribution files to PyPI using twine
+   (``twine upload -s dist/*``). The upload should include GPG signatures of
+   the distribution files.
+#. Push all of the changes.
+#. Regenerate the ``get-pip.py`` script by running
+   ``invoke generate.installer`` in the get-pip repository, and committing the
+   results.
+
+Doing a bug-fix release
+-----------------------
+
+Sometimes we need to release a bugfix release of the form ``X.Y.Z+1``. In order
```

Comment:
With date-based releases, will we still do bugfix releases? If we do, what's the version numbering? I understood the basic numbering to be `18.1`, `18.2`, ... (or do we start at 0? There's a few milestones/issues referring to 18.0) - and if so, do we have a bugfix release for `18.1` being `18.1.1` (meaning we don't have a consistent number of parts) or do we do `18.1.0` as the initial release (which I don't like, because it looks too like major.minor.patch rather than year.sequence).

Reply:
From @dstufft's original PR, #5324: 

> Given our release cadence, that means that (pretending we had used this scheme all along) the versions for 2018 would look like:
>
> | Month | Version |
> | --- | --- |
> | Jan | 18.0 |
> | Apr | 18.1 |
> | Jul | 18.2 |
> | Oct | 18.3 |
>
> [snip]
>
> If we need to make any bug fixes to an already released version, then we would add a bug fix version number, like 18.0.1, 18.0.2, etc.
>
> Since we're starting part way through the year, our first release with this scheme would be in July, and would be 18.0.